File: go/0977E-有序数组的平方.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-09 19:32:50
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func sortedSquares(nums []int) []int {
	n := len(nums)
	ans := make([]int, n)
	i, j := 0, n-1
	for k := n - 1; k >= 0; k-- {
		powI, powJ := nums[i]*nums[i], nums[j]*nums[j]
		if powI > powJ {
			ans[k] = powI
			i++
		} else {
			ans[k] = powJ
			j--
		}
	}
	return ans
}

File: cpp/0977E-有序数组的平方.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-13 09:12:57
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

#include <bits/stdc++.h>

using namespace std;

class Solution {
   public:
    vector<int> sortedSquares(vector<int>& nums) {
        vector<int> ans(nums.size(), 0);
        for (int i = 0, j = nums.size() - 1, k = j; i <= j;) {
            int ii = nums[i] * nums[i];
            int jj = nums[j] * nums[j];
            if (ii > jj) {
                ans[k--] = ii;
                i++;
            } else {
                ans[k--] = jj;
                j--;
            }
        }

        return ans;
    }
};